consensus/src/runway/mod.rs

```diff
@@ -521,9 +526,15 @@ where
                 self.store.add_parents(h, p_hashes);
                 self.resolve_missing_parents(&h);
                 if let Some(su) = self.store.unit_by_hash(&h).cloned() {
+                    self.send_message_for_network(RunwayNotificationOut::NewAnyUnit(
```

Comment:
I have thought about it for quite a bit and it now seems to me that this should be triggered not on `AddedToDag` but on each call to `self.store.add_unit`. The difference is that `AddedToDag` only happens after we have complete "history" of the unit, i.e. its parents and their parents, etc. so all its "lower cone". But `self.store.add_unit` is called whenever we get a new, valid unit. There is no reason why we should not rebroadcast units without full history.

For this reason I would like you to move sending this message from here to the (3 ?) places where `self.store.add_unit(_, _)` is called and send it from there. I think the best approach would be add a `bool` return value for `add_unit` to return a `bool` of whether the unit was added or was a duplicate (so that we can ignore duplicates).

Reply:
I would be wary of this approach, before a unit is properly inserted in the DAG it might be malicious in some way, and rebroadcasting it then might not be a good idea. In practice units get inserted quickly enough that this shouldn't matter for performance reasons, right?

Reply:
For malicious units it shouldn't matter much, I think.
But my impression was that if we rebroadcast only DAG units then we violate the `ES` property even more... Maybe that's not true?

Reply:
That would be quite strange, but maybe because I very much operate on the assumption that only in-DAG units are actually important (clearly for the consensus, but I'm not sure about other things). I cannot think of a specific attack with a malicious unit (create a unit of round 2137, and afterwards no one broadcasts your actual units?), but dropping that assumption seems strange to me.

Reply:
Ok, I'm convinced. I think that by "induction" one should be able to argue that the set of all "honest" units will eventually be disseminated, because let's say `r` is the lowest round on which there is a unit `U` which has issues with dissemination, but then all units below it (having round `<=r-1`) have been disseminated correctly, hence they are in DAG, hence `U` itself in the DAG of the honest node holding it, and thus will be rebroadcast. 

@obrok -- you can ignore my above comment then, the current version is OK :P I'm leaving this discussion here. Feel free to resolve.

Reply:
Heh, I already changed it, will revert.